File: typescript/src/core/token.ts
/**
 * An immutable token carrying a typed value through the Petri net.
 *
 * Tokens flow from place to place as transitions fire, carrying typed
 * payloads that represent the state of a computation or workflow.
 */
export interface Token<T> {
  readonly value: T;
  /** Epoch milliseconds when the token was created. */
  readonly createdAt: number;
  /**
   * JSON-friendly projection of the token value, populated by
   * {@link SessionArchiveReader} when hydrating a v3 archive so replay
   * consumers see the same structured shape the writer emitted. Live tokens
   * (produced by {@link tokenOf} / {@link tokenAt}) leave this `undefined`;
   * the runtime ignores it. See [EVT-025](../../../spec/08-events-observability.md)
   * AC5. (libpetri 1.8.0+)
   */
  readonly structured?: unknown;
}

/** Cached singleton unit token. */
const UNIT_TOKEN: Token<null> = Object.freeze({
  value: null,
  createdAt: 0,
});

/**
 * Creates a token with the given value and the current wall-clock timestamp.
 *
 * **Reads the real clock, always.** It is called from host code, so no executor seam reaches it
 * — an injected {@link import('../runtime/clock.js').Clock} does not change what it stamps
 * ([TIME-015]). It is therefore non-deterministic under a host clock and **unsuitable for an
 * initial marking**, whose tokens are minted before any executor exists: a replay reseeded with
 * `tokenOf` differs from the previous attempt inside the marking itself. Use
 * {@link import('../runtime/clock.js').seedToken} for fresh seed tokens, or {@link tokenAt} to
 * state the timestamp outright.
 */
export function tokenOf<T>(value: T): Token<T> {
  return { value, createdAt: Date.now() };
}

/**
 * Returns a unit token (marker with no meaningful value).
 * Used for pure control flow where presence matters but data doesn't.
 * Returns a cached singleton whose `value` is `null`.
 */
export function unitToken(): Token<null> {
  return UNIT_TOKEN;
}

/**
 * Creates a token with a specific timestamp (for testing/replay).
 *
 * The explicit-timestamp escape hatch ([CORE-011]) behind every determinism rule here, and what
 * {@link import('../runtime/clock.js').seedToken} is built from.
 *
 * Not needed for ordinary replay-safety: under an injected clock libpetri already stamps every
 * token it constructs — `ctx.output(...)` included — from that clock ([TIME-015] AC#13). Reach
 * for this when the host wants to *choose* a timestamp: replaying a recorded token, restoring a
 * snapshot ([CORE-073]), or seeding a marking before any executor exists.
 */
export function tokenAt<T>(value: T, createdAt: number): Token<T> {
  return { value, createdAt };
}

/** Checks if this is the singleton unit token. */
export function isUnit(token: Token<unknown>): boolean {
  return token === UNIT_TOKEN;
}

File: java/src/main/java/org/libpetri/core/Token.java
package org.libpetri.core;

import java.time.Instant;

/**
 * An immutable token carrying a typed value through the Petri net.
 * <p>
 * Tokens are the "data carriers" of the net. They flow from place to place
 * as transitions fire, carrying typed payloads that represent the state of
 * a computation or workflow.
 * <p>
 * Tokens are immutable and include a creation timestamp for timing analysis
 * and debugging.
 *
 * <h3>Example</h3>
 * <pre>{@code
 * Token<Request> token = Token.of(new Request("search", "laptop"));
 * Request value = token.value();
 * Instant created = token.createdAt();
 * }</pre>
 *
 * <h3>Unit Tokens</h3>
 * For control flow where no data is needed, use {@link #unit()} which returns
 * a cached singleton to avoid allocation.
 *
 * @param <T> the type of value this token carries
 * @param value the typed payload (may be null for unit tokens)
 * @param createdAt timestamp when the token was created
 */
public record Token<T>(
    T value,
    Instant createdAt
) {
    private static final Token<Void> UNIT = new Token<>(null, Instant.EPOCH);

    /**
     * Creates a token with the given value and current timestamp.
     *
     * @param <T> the type of the value
     * @param value the payload to carry
     * @return a new token with the value and current time
     */
    public static <T> Token<T> of(T value) {
        return new Token<>(value, Instant.now());
    }

    /**
     * Returns a unit token (marker with no meaningful value).
     * <p>
     * Unit tokens are used for pure control flow where the presence of
     * a token matters but it carries no data. Returns a cached singleton.
     *
     * @param <T> the phantom type (allows unit tokens in any typed place)
     * @return the cached unit token singleton
     */
    @SuppressWarnings("unchecked")
    public static <T> Token<T> unit() {
        return (Token<T>) UNIT;
    }

    /**
     * Checks if this token's value is an instance of the given type.
     *
     * @param type the class to check against
     * @return true if value is non-null and is an instance of type
     */
    public boolean isType(Class<?> type) {
        return value != null && type.isInstance(value);
    }

    /**
     * Checks if this is the unit token (no meaningful value).
     *
     * @return true if this is the singleton unit token
     */
    public boolean isUnit() {
        return this == UNIT;
    }
}
